File: frontend/app/src/main/scala/app/domain/followers/Group.scala
package app.domain.followers

import app.domain.core.{AbstractEntity, Data, Link, Table, User}
import app.support.Api
import app.support.Api.given
import jfx.core.meta.Meta
import jfx.core.state.{ListProperty, Property}
import jfx.form.validators.*

import java.util.UUID
import scala.concurrent.Future
import scala.scalajs.js
import scala.scalajs.js.URIUtils.encodeURIComponent

class Group extends AbstractEntity {

  @NotBlank(message = "Name ist erforderlich")
  @Size(min = 3, max = 80, message = "Name muss zwischen 3 und 80 Zeichen haben")
  val name: Property[String] = Property("")
  val users: ListProperty[User] = ListProperty()

  val editable: Property[Boolean] = Property(false)

  def save(): Future[Data[Group]] =
    Api.request("/service/followers/groups/groups").post(this).read[Data[Group]]

  def update(): Future[Data[Group]] =
    Api.request("/service/followers/groups/groups").put(this).read[Data[Group]]

  def delete(): Future[Unit] =
    Api.request("/service/followers/groups/groups").delete(this).unit
}

object Group {
 
  def read(id: String): Future[Data[Group]] =
    Api.request(s"/service/followers/groups/groups/$id").get.read[Data[Group]]

  def list(index: Int, limit: Int, sorting: Seq[String] = Seq("created:desc")): Future[Table[Data[Group]]] =
    Api.request(s"/service/followers/groups?index=$index&limit=$limit${renderSortParameters(sorting)}").get.read[Table[Data[Group]]]

  private def renderSortParameters(sorting: Seq[String]): String = {
    val normalizedSorting = sorting.iterator.map(_.trim).filter(_.nonEmpty).toVector
    if (normalizedSorting.isEmpty) ""
    else normalizedSorting.map(value => s"&sort=${encodeURIComponent(value)}").mkString
  }
}

File: frontend/app/src/main/kotlin/app/domain/followers/Group.kt
package app.domain.followers

import app.domain.core.AbstractEntity
import app.domain.core.Address
import app.domain.core.Data
import app.domain.core.Email
import app.domain.core.Link
import app.domain.core.Media
import app.domain.core.Table
import app.domain.core.User
import app.domain.core.UserInfo
import jFx2.client.JsonClient
import jFx2.state.ListProperty
import jFx2.state.ListPropertySerializer
import jFx2.state.Property
import jFx2.state.PropertySerializer
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlin.time.Clock

@Serializable
data class Group(
    @Serializable(with = PropertySerializer::class)
    override var id : Property<String>? = null,
    @Serializable(with = PropertySerializer::class)
    val name : Property<String> = Property(""),
    @Serializable(with = PropertySerializer::class)
    override val modified : Property<LocalDateTime> = Property(Clock.System.now().toLocalDateTime(TimeZone.UTC)),
    @Serializable(with = PropertySerializer::class)
    override val created : Property<LocalDateTime> = Property(Clock.System.now().toLocalDateTime(TimeZone.UTC)),
    @Serializable(with = ListPropertySerializer::class)
    val users : ListProperty<User> = ListProperty(),
    @SerialName($$"$links")
    @Serializable(with = ListPropertySerializer::class)
    override val links : ListProperty<Link> = ListProperty()
) : AbstractEntity {

    @Transient
    val editable = Property(false)

    suspend fun save() : Data<Group> {
        return JsonClient.post("/service/followers/groups/groups", this)
    }

    suspend fun update() : Data<Group> {
        return JsonClient.put("/service/followers/groups/groups", this)
    }

    suspend fun delete() {
        JsonClient.delete("/service/followers/groups/groups", this)
    }

    companion object {

        suspend fun read(id : String) : Data<Group> {
            return JsonClient.invoke<Data<Group>>("/service/followers/groups/groups/$id")
        }

        suspend fun list(index : Int, limit : Int) : Table<Data<Group>> {
            return JsonClient.invoke<Table<Data<Group>>>("/service/followers/groups?index=$index&limit=$limit&sort=created:desc")
        }
    }


}
